File: KWAudioUnit/KWAudioUnit/FRAudioUnit/FROutputUnit.m
//
//  FROutputUnit.m
//  TestCase_AudioController
//
//  Created by 상우 강 on 2016. 5. 27..
//  Copyright © 2016년 상우 강. All rights reserved.
//

#import "FROutputUnit.h"

@implementation FROutputUnit

-(id)init {
    if (self = [super initWithType:kFRAudioUnitType_Output]) {

    }
    return self;
}

-(void)initializeAudioUnit {
    [super initializeAudioUnit];
    AudioUnit audioUnit = [self audioUnit];
    OSStatus err = AudioOutputUnitStart(audioUnit);
    if (err)
        NSLog(@"outputUnit start failed : %d", (int)err);
}

-(void)uninitialize {
    OSStatus err = noErr;
    err = AudioOutputUnitStop(self.audioUnit);
    if (err)
        NSLog(@"output unit stop failed : %d", (int)err);
}

-(void)setOutputVolume:(float)volume {
    OSStatus err;
    err = AudioUnitSetParameter(self.audioUnit, kHALOutputParam_Volume, kAudioUnitScope_Output, 0, volume, 0);
    if (err)
        NSLog(@"audio unit set param for output volume : %d", (int)err);
}

@end

File: Testing_iOS/Testing_iOS/lib/FROutputUnit.swift
//
//  FRIOUnit.swift
//  FRAudioUnit
//
//  Created by exs-mobile 강상우 on 02/08/2019.
//  Copyright © 2019 강상우. All rights reserved.
//

import UIKit
import AudioToolbox

class FROutputUnit: FRAudioUnit {
    
    init() {
        super.init(.IO)
    }
    
    override func initializeAudioUnit() {
        super.initializeAudioUnit()
        
        var err = noErr
        err = AudioOutputUnitStart(self.audioUnit!)
        if err != noErr {
            printError("io unit start error", err)
        }
    }
    
    override func unInitializeAudioUnit() {
        var err = noErr
        err = AudioOutputUnitStop(self.audioUnit!)
        if err != noErr {
            printError("io unit stop error", err)
        }
        
        super.unInitializeAudioUnit()
    }
    
    //MARK: - Public
    public func setOutputVolume(volume: Float) {
        var err = noErr
        err = AudioUnitSetParameter(self.audioUnit!, kHALOutputParam_Volume, kAudioUnitScope_Output, 0, volume, 0)
        if err != noErr {
            printError("output volue set", err)
        }
    }
}
